### PR TITLE
backtrack if can not activate

### PR DESCRIPTION
**Pull request by Eh2406**
This is a fix for #4347
Unfortunately this too regressed error messages for the case that you specified a dependency feature that does not  exist. 
@alexcrichton advice on improving the message?

### CODE REVIEW

**Comment by alexcrichton** [comment]
Oh dear that's a bummer! It seems like a lot of roads are pointing towards error messages not being that great...

This one I would say though yeah is even more pressing than `links` in that this comes up quite frequently (typos and such). In that sense it'd be pretty bad if we regressed this error message :(

**Comment by Eh2406** [comment]
Is there some way to print a error just as we would have but then continue
processing, if so I have a bunch of ideas for how to improve error
messages, at least in this case.

Or is there some way to tie an an error message to a previous one. So we
can say options for dependent b causes include...

**Comment by Eh2406** [comment]
It will take some work to rebase/merge this with #4834, and that is a much bigger/more common UX improvement. I think this should be put on hold for that to land and for a plan for error messages to be developed.
@alexcrichton You are right, I have been thoroughly convinced that error messages should happen next. You offered some mentership on that, how do you think I should proceed?

**Comment by alexcrichton** [comment]
@Eh2406 certainly! Unfortunately I don't really know how to precisely proceed. I do have a feeling, though, for how to proceed generally!

So prior to https://github.com/rust-lang/cargo/pull/4978 we had a custom error message for the `links` key which indicated. I think that a great end state would be to end up basically as close as we can to an error message like that where you've got a causal chain through packages and a nice description for what's happening.

As to how to get there... I'm not sure. I think it may also be worth taking a step back and looking at the resolution algorithm from a distance too. Right now Cargo in theory attempts all possible crate resolution graphs. I mean like *all possible graphs*. This can put us in one of two situations:

* At least one graph is "successful" and satisfies all the constraints. In this situation Cargo should never return an error and eventually reach such a graph.
* All possible graphs would lead to a resolution failure, so Cargo should return an error.

Here we're most interested in the latter situation, but it also means there's no one error for us to return. In some sense we tried everything and it could have all failed for a number of reasons. I think we could probably optimize to return the "shortest" error perhaps? (for whatever definition "shortest" actually means)

There's also the separate bug that Cargos' backtracking and exploration of the problem space takes forever, but hopefully https://github.com/rust-lang/cargo/pull/4834 can help with that!

---

Hm so that's unfortunately not super actionable, but is that helpful at least to get started?

**Comment by Eh2406** [comment]
Thanks for the pointer to the `links` code, I will go and read that. I like the idea of having a good causal chain for the cases when we can.

I think it may be useful to break it down into more situations. 

- The tree is trivially solvable, even if it is big. This is common, For example Servo, when I tested it, does No backtracking at all. -> No error messages. 
- The tree is solvable, given a exponential amount of time. #4834 will make it much more rare, But the problem is NP complete so this is always a possibility. -> We should print some kind of diagnostic to help the programer make it easier for us. Ether at some time into resolution like we do [Resolving dependency graph...](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/mod.rs#L650) Or wen we bail after a user configured number of time as you suggested in #4066
- The tree has no solutions. -> We need to report something useful to the programer.

We should keep in mind that the most common reason for any kind of error is a simple typo.
And that the user's mental model does not include millions of permutations of ancient versions of the dependencies.

The first time we discover that the tree is not trivial we are in the state closest to the user's mental model of just take the newest version of everything. And almost anything we print will be helpful in discovering that it is just a typo.  So I think we should print a good causal chain for the backtracking to the user, and continue in the serch.

Another huristick if something deep in our tree is unsolvable, then we probably hit almost the same error message lots of times. For example We depend on A=* (with 100 versions to choose from) which depends on B=* (with 100 versions to choose from) which depends on C=* all versions have been yanked, If we printed all errors (a thout experiment) We will have 10k "can't find a candidate for C ...." and just 100 "can't find a candidate for B ...." and 1 "can't find a candidate for A ...." The most common one is the most helpful. So maybe 10 minutes into resolution we can print "we are still working on finding something that works but, here are the 3 most often encountered difficulties:" each with a full a good causal chain for the backtracking.

All of these thoughts depend on being able to print the message as part of some context, and then continue with the serch. How do I do that? I could make a function that returns the errors as strings. But the words "errors as strings" makes me very nervous.

**Comment by alexcrichton** [comment]
All excellent points! I think we should take it as an assumption that *if* we're going to produce a human readable error then we can reach that conclusion in a reasonable amount of time. In that sense I hope that #4834 will get us closer to that goal but we may have other bugs of "Cargo goes in the weeds too much". I don't think, though, that such a situation should impact the design of the error messages here.

Now I will say we have two other tricks up our sleeve for "gee resolution is taking forever":

* First we can actually inform the user what's happening. Recently Cargo changed to do this and will print out, after I think a second or so of resolution, "Resolving crate graph...". It's not much but it at least let's the user know we're not stuck in network traffic.
* Second we should probably implement Cargo just giving up at some point. If Cargo has attempted a million graphs and they're all failing, it's doubtful that the million+1th graph will work. This is sort of covered by https://github.com/rust-lang/cargo/issues/4066 where we should (a) have a hard limit on the amount of time Cargo takes during resolution and (b) when this error happens mention something like "hey this may be a bug in Cargo, just letting you know"

In that sense I think your second and third cases would fold into one another. I definitely agree that mistakes tend to be common typos before we get to serious "oh dear what to do" scenarios. Despite that though if Cargo is indeed successful I don't think that we should print out anything (even if we hit errors along the way). I think that only if we end up failing entirely should we print an error.

We could certainly explore though a more progress-bar like situation for Cargo where it prints incrementally what it's doing for debugging perhaps?

> All of these thoughts depend on being able to print the message as part of some context, and then continue with the serch. How do I do that? I could make a function that returns the errors as strings. But the words "errors as strings" makes me very nervous.

I think it's sort of encoded as `CargoResult` or weird options in resolve today, but I think we should stick with that. An error happening only happens in the "slow" case where "slow" means that you've modified or don't already have a Cargo.lock. On the fast path we'll never hit an error because the Cargo.lock means we'll resolve precisely to the same one as before.

In that sense I wouldn't be too worried about making errors expensive to compute, it hopefully isn't too bad! We can of course profile and test on Servo though to test this out.

**Comment by Eh2406** [comment]
Ok, I think we mostly agree and if we don't it is because we are talking past each other. I think I need to start thinking more "incrementally". I should start with reading the old 'links' code and seeing if I can make the current error messages better. Only then come back to the larger questions of when to print and for witch backtracks.

File: tests/testsuite/features.rs [hunk 1]
```diff
@@ -109,8 +109,17 @@ fn invalid4() {
 
     assert_that(p.cargo("build"),
                 execs().with_status(101).with_stderr("\
-[ERROR] Package `bar v0.0.1 ([..])` does not have these features: `bar`
-"));
+error: failed to select a version for `bar`.
+    ... required by package `foo v0.0.1 ([..])`
+versions that meet the requirements `*` are: 0.0.1
+
+the package `bar` depends on `bar`, with features: `bar` but it does not have these features.
```
Comment (alexcrichton) [inline tests/testsuite/features.rs]: Hm shouldn't this read "the package foo" instead of "the package bar"?

Comment (Eh2406) [reply tests/testsuite/features.rs]: Yes, and it is fixed. Thanks. (Sorry about all the force pushing.)

**Comment by Eh2406** [comment]
I think this is finally ready to go.

Some things I did not figure out how to test:
1.  Where to clone BacktrackFrame. I think I have this right. I think my initial version could have had `O(n^2)`, but I could not get a test case that proved it.
2. Adding conflicting_activations to the BacktrackFrame. I think this is needed as I described in the commit, but I could not get a test case that proved it.

File: tests/testsuite/features.rs [hunk 2]
```diff
@@ -109,8 +109,17 @@ fn invalid4() {
...
+package `bar v0.0.1 ([..])`
+    ... which is depended on by `foo v0.0.1 ([..])`
+
+all possible versions conflict with previously selected packages.
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply tests/testsuite/features.rs]: Removed

**Comment by alexcrichton** [comment]
Nah this looks great to me, thanks @Eh2406 

@bors: r+

**Pull request merged**